File: 2871-double-a-number-represented-as-a-linked-list/double-a-number-represented-as-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def doubleIt(self, head: Optional[ListNode]) -> Optional[ListNode]:
        carry = [0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1]
        nums = [0,1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6,7,8]

        dummy = ListNode(0)
        dummy.next = head
        prev = dummy
        curr = head

        while curr:
            val = curr.val << 1
            prev.val += carry[val]
            curr.val = nums[val]
            prev = curr
            curr = curr.next
        
        return head if dummy.val == 0 else dummy
        

File: 2871-double-a-number-represented-as-a-linked-list/double-a-number-represented-as-a-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode doubleIt(ListNode head) {
        if (head == null){
            return head;
        }
        Stack<ListNode> st = new Stack<ListNode>();

        ListNode p = head;
        while (p!=null){
            ListNode q = p;
            p = p.next;
            st.push(q);
        }

        ListNode prev = null;
        int carry = 0;

        while (st.size() != 0){
            ListNode q = st.pop();
            int val = q.val;
            val *= 2;
            val += carry;
            q.val = val % 10;
            carry = val / 10;
            q.next = prev;
            prev = q;
        }

        if(carry > 0){
            ListNode q = new ListNode(carry);
            q.next = prev;
            prev = q;
        }

        return prev;

    }
}